src/SFML/Graphics/RenderTarget.cpp

```diff
@@ -317,6 +319,107 @@ void RenderTarget::draw(const Vertex* vertices, std::size_t vertexCount,
 
 
 ////////////////////////////////////////////////////////////
+void RenderTarget::draw(const VertexBuffer& vertexBuffer, const RenderStates& states)
+{
+    draw(vertexBuffer, 0, vertexBuffer.m_size, states);
+}
+
+
+////////////////////////////////////////////////////////////
+void RenderTarget::draw(const VertexBuffer& vertexBuffer, std::size_t firstVertex, std::size_t vertexCount, const RenderStates& states)
+{
+    // VertexBuffer not supported?
+    if (!VertexBuffer::isAvailable())
+    {
+        err() << "sf::VertexBuffer is not available, drawing skipped" << std::endl;
+        return;
+    }
+
+    // Sanity check
+    if (firstVertex > vertexBuffer.m_size)
+        return;
+
+    // Clamp vertexCount to something that makes sense
+    vertexCount = std::min(vertexCount, vertexBuffer.m_size - firstVertex);
+
+    // Nothing to draw?
+    if (!vertexCount || !vertexBuffer.m_buffer)
+        return;
+
+    // GL_QUADS is unavailable on OpenGL ES
+    #ifdef SFML_OPENGL_ES
+        if (vertexBuffer.m_primitiveType == Quads)
+        {
+            err() << "sf::Quads primitive type is not supported on OpenGL ES platforms, drawing skipped" << std::endl;
+            return;
+        }
+        #define GL_QUADS 0
+    #endif
+
+    if (setActive(true))
+    {
+        // First set the persistent OpenGL states if it's the very first call
+        if (!m_cache.glStatesSet)
+            resetGLStates();
+
+        applyTransform(states.transform);
+
+        // Apply the view
+        if (m_cache.viewChanged)
+            applyCurrentView();
+
+        // Apply the blend mode
+        if (states.blendMode != m_cache.lastBlendMode)
+            applyBlendMode(states.blendMode);
+
+        // Apply the texture
+        Uint64 textureId = states.texture ? states.texture->m_cacheId : 0;
+        if (textureId != m_cache.lastTextureId)
+            applyTexture(states.texture);
+
+        // Apply the shader
+        if (states.shader)
+            applyShader(states.shader);
+
+        // Bind vertex buffer
+        VertexBuffer::bind(&vertexBuffer);
+
+        // Always enable texture coordinates
+        if (!m_cache.texCoordsArrayEnabled)
+            glCheck(glEnableClientState(GL_TEXTURE_COORD_ARRAY));
+
+        glCheck(glVertexPointer(2, GL_FLOAT, sizeof(Vertex), reinterpret_cast<const void*>(0)));
+        glCheck(glColorPointer(4, GL_UNSIGNED_BYTE, sizeof(Vertex), reinterpret_cast<const void*>(8)));
+        glCheck(glTexCoordPointer(2, GL_FLOAT, sizeof(Vertex), reinterpret_cast<const void*>(12)));
```

Comment:
Can't this be cached? This only needs to be called when a `VertexArray` was draw previously, which should happen less and less, and not anymore in SFML 3.

Reply:
From the documentation:

> If a non-zero named buffer object is bound to the GL_ARRAY_BUFFER target (see glBindBuffer) while a vertex array is specified, pointer is treated as a byte offset into the buffer object's data store. Also, the buffer object binding (GL_ARRAY_BUFFER_BINDING) is saved as vertex array client-side state (GL_VERTEX_ARRAY_BUFFER_BINDING).

The buffer binding is saved as part of the array state. If you draw from another buffer, you will have to rebind the pointers since GL_ARRAY_BUFFER_BINDING changed. This is the weird kind of legacy emulation the driver has to pull off because we aren't making use of VAOs.

Reply:
Ok, I didn't know (and expect) this kind of behavior.